### DISCUSSION TITLE
Breaking change: DateTimeField validation parsing

### DISCUSSION THREAD

**Discussion by RignonNoel** [encode/django-rest-framework#9156] [Potential Issue]
We upgraded a Django project to the new version of Django (`4.2.5`) and DRF (`3.14.0`) and some of ours unit test failed due to breaking changes on the `DateTimeField()` parsing validation. 

_Example of serializer:_
```
class MySerializer(serializers.Serializer):
    start_time = serializers.DateTimeField()
```
_Example of data we tested:_ 
```
2021-03-13
```

_Example of exception we expected before:_
```
Datetime has wrong format. Use one of these formats instead: YYYY-MM-DDThh:mm[:ss[.uuuuuu]][+HH:MM|-HH:MM|Z].
```

**Now, it seems that `2021-03-13` is valid... But it's not at all matching the format proposed in the default error message that DRF use: `YYYY-MM-DDThh:mm[:ss[.uuuuuu]][+HH:MM|-HH:MM|Z].`**

---

Different option i think:

1. We adapt the default error message. But for me it's a breaking change since a lot of platform will have datetime with a default `00:00:00` time on datetime in place of the error message they got before
2. We include a date format validation inside DRF to match the format we had before
3. We cooperate with Django repo to adapt the change they made since this change could have a lot of other repercution on a lot of other platform (making the `parse_datetime()` more clearly defined to avoid futur problem)

---

## Here are my research

https://github.com/encode/django-rest-framework/blob/f56b85b7dd7e4f786e0769bba6b7609d4507da83/rest_framework/settings.py#L107

https://github.com/encode/django-rest-framework/blob/f56b85b7dd7e4f786e0769bba6b7609d4507da83/rest_framework/fields.py#L1177-L1198

We see here that DRF clearly use `ISO_8601` in the default settings and use the django `parse_datetime()` function to make the logic attached

https://github.com/django/django/blame/656192c2c96bb955a399d92f381e38fe2254fe17/django/utils/dateparse.py#L104-L130

But we see here that Django try to use `datetime.fromisoformat()` first since this [commit](https://github.com/django/django/commit/f35ab74752adb37138112657c1bc8b91f50e799b) of @kezabelle and `fromisoformat()` allow a LOT MORE format than before as you can see on their documentation

![Firefox_Screenshot_2023-11-03T18-52-46 634Z](https://github.com/encode/django-rest-framework/assets/12053720/8181033c-da60-469e-9965-7bcec46414a7)

https://docs.python.org/3/library/datetime.html#datetime.datetime.fromisoformat

**Comment by auvipy** [comment]
I am open to see a PR with a failing test case atleast

**Reply by RignonNoel** [reply]
The problem is really straightforward to test since it's the default validor of any `DateTimeField`.

But yeah, I will open a PR since we didn't got any news in the last two weeks :)

**Reply by RignonNoel** [reply]
@auvipy I don't have any open-source repository with this problem right now on hand, and I don't have the time to create a full new repo of proof just for that. But please feel free to let me know what would be the best way to give you a better test if it's too complex for you to test the default validator by yourself. 

Basically if you create any serialiser as simple as this one:

```
class MySerializer(serializers.Serializer):
    start_time = serializers.DateTimeField()
```

And that you give the value `2021-03-13` in input for validation (ie: `MySerializer(data).is_valid()`), you will get the problem I explained.

**Reply by auvipy** [reply]
Thanks. It seems to be a Django compat issue and I will look into it much further

**Reply by DNiemeijer** [reply]
I am running into the exact same issue as described above during a Django upgrade. Do you have an indication on when we can expect an update on this? Thanks in advance :)

**Reply by RignonNoel** [reply]
@DNiemeijer An issue has been opened here if you want to follow it also: https://github.com/encode/django-rest-framework/issues/9161

**Reply by auvipy** [reply]
I would be happy to see any fix with minimal test case
